### PR TITLE
read write lock

### PR DESCRIPTION
**Pull request by janbar**
No mutex: the speed is 3 x faster, and no need pthread extra.

### CODE REVIEW

**Comment by janbar** [comment]
it is a sample. The final implementation should not use std::unique_lock or std::shared_lock. So we have to make own wrappers with functions lock() , unlock() and lock_shared(), unlock_shared().

**Comment by janbar** [comment]
The last push implements the ReadLock and WriteLock wrapper, with same way of std:shared_lock en std::unique_lock.
Obviously, like the stl wrappers, we cannot exchange those wrappers between thread. The code in libosmscout check that, so no issue.

**Comment by Karry** [comment]
Well, how own locks are better than STL ones? `std::shared_lock<SharedMutex>` and `std::unique_lock<SharedMutex>`... I do not see any benefit of them. I would prefer standard implementation with less own code to maintain.

In general, I like your implementation without pthread, but it is spin-lock in user-space with all downsides. When writer will be blocked by IO operation for some time and there will be some reader threads waiting for the lock, they will be burning cpu cycles in active waiting loop...

**Comment by janbar** [comment]
Even, testing I found few cases of deadlock: Some readers hold two or three times the shared lock. When a writer take priority, it is waiting for release of the shared. But those readers are locked trying to lock again.
So it will be nice to find a way avoiding the use of something not implemented by all platforms. I will try to think about this using 2 mutex to make a real LATCH, and I will let you know.

**Comment by Karry** [comment]
You know, previously DBThread was using QReadWriteLock from Qt. It has writer precedence. But with continuous process to move some basic blocks from cilent-qt library to Qt-less client library, we cannot use this lock in DBThread anymore... I was not aware of the fact that STL implementation behaves differently than Qt one :-/

**Comment by janbar** [comment]
I guess the issue is the usage of "reentrant mutex", and that cause the deadlock. The std::shared_mutex has writer precedence (have a look in the libstd++ implementation), but you have to care of the readers don't lock more one times. Fixing that is not impossible and even could clarify the code. I could help to fix it, and then we can use the standard std::shared_mutex.

**Comment by janbar** [comment]
I succeeded to resolve the deadlock. It requires some changes but not too much. I will push my change soon.

**Comment by janbar** [comment]
The idea is:
- Reader shouldn't lock more than 1 times (non reentrant), obviously a reader cannot lock for shared mode, then lock for exclusive mode.
- A writer locks for exclusive mode, and then it could lock for shared mode too.

Underwood, std:shared_mutex uses a counter for shared mode, and a safe flag protected with mutex. In some cases it could use pthread facilities.

**Comment by janbar** [comment]
I will make an other version supporting reentrant lock for testing. This requires to store an array in Latch containing the reader thread ids that own the shared lock.

**Comment by janbar** [comment]
Voila !
Now we have a Latch fully re-entrant. And finally no more deadlock, the writers have precedence, and DBLoadJob no longer crashes. I use a spin-lock only for internal access to the s_lockers list. The thread blocking uses wait/signal, so no longer a spin-lock, and only when necessary:

readers/readers => no locking, spin-lock to share s_locker list.
readers/writer => readers are locked and wait for x_gate.
readers/writer1/reader => writer1 has set x_flag, and is locked and wait for s_gate, readers are locked and wait for x_gate.

**Comment by janbar** [comment]
few bugs fixed, now I am running benchmark on the last push. I will let you know...

**Comment by janbar** [comment]
few bugs fixed, now I am running benchmark on the last push. I will let you know...

The implementation by pthread doesn't allow a reader to be upgraded as writer, also double unlock is not safe and will throw. Same with std::shared_mutex. My implementation allows these features, and it why it wastes some time in system calls (alloc/free from std::list):

- pthread:
real	0m2,808s
user	0m18,778s
sys	0m0,326s

- mine:
real	0m5,811s
user	0m17,920s
sys	0m16,080s  >>> alloc/free of struct s_locker

I keep left this version for me, and I remove those features to be compared.

**Comment by janbar** [comment]
So, I updated the source to remove reentrant reader/writer that required to maintain the shared ids, and now it is comparable i.e same features than pthread. The benchmark is following:

10 readers are executing 2.000.000 of reads each ,and  2 writers are execution 200.000 of increments on the counter each.
One status line is printed every 100ms.

- pthread:
> time ./build/readwritelock2 
Running (12): counter=0 , readers: ?
Running (12): counter=0 , readers: ?
Running (12): counter=0 , readers: ?
Running (12): counter=0 , readers: ?
Running (12): counter=1 , readers: ?
Running (12): counter=1 , readers: ?
Running (12): counter=1 , readers: ?
Running (12): counter=1 , readers: ?
Running (12): counter=1 , readers: ?
Running (12): counter=1 , readers: ?
Running (12): counter=1 , readers: ?
Running (12): counter=1 , readers: ?
Running (12): counter=1 , readers: ?
Running (12): counter=1 , readers: ?
Running (12): counter=1 , readers: ?
Running (12): counter=1 , readers: ?
Running (12): counter=1 , readers: ?
Running (12): counter=1 , readers: ?
Running (12): counter=1 , readers: ?
Running (12): counter=1 , readers: ?
Running (11): counter=2 , readers: ?
Running (10): counter=2 , readers: ?
Running (10): counter=222 , readers: ?
Running ( 7): counter=9041 , readers: ?
Running ( 4): counter=28950 , readers: ?
Running ( 3): counter=49666 , readers: ?
Running ( 3): counter=70305 , readers: ?
Running ( 3): counter=400000 , readers: ?
Completed   : counter=400000

**real	0m2,809s**
user	0m17,793s
sys	0m0,432s

- mine:
> time ./build/readwritelock
Running (12): counter=0 , readers: 1
Running (12): counter=13171 , readers: 1
Running (12): counter=32407 , readers: 0
Running (12): counter=52162 , readers: 2
Running (12): counter=76455 , readers: 0
Running (12): counter=97463 , readers: 0
Running (12): counter=117066 , readers: 0
Running (12): counter=137522 , readers: 2
Running (12): counter=161012 , readers: 0
Running (12): counter=184842 , readers: 0
Running (12): counter=206932 , readers: 3
Running (12): counter=231479 , readers: 1
Running (12): counter=248991 , readers: 0
Running (10): counter=266539 , readers: 0
Running ( 8): counter=353887 , readers: 0
Running ( 7): counter=400000 , readers: 0
Completed   : counter=400000

**real	0m1,609s**
user	0m3,051s
sys	0m2,748s

And one remark, pthread seems to be satured on start and writers have not really the priority in this situation.
My latch is never satured and really give the priority to the writers.

File: libosmscout/include/osmscout/async/ReadWriteLock.h [hunk 1]
```diff
@@ -24,48 +24,189 @@
...
-    SharedMutex& operator=(const SharedMutex&) = delete;
-    SharedMutex& operator=(SharedMutex&&) = delete;
-    ~SharedMutex();
-
-    void lock();
-    void unlock();
-
-    void lock_shared();
-    void unlock_shared();
+class Latch {
```
Comment (Karry) [inline libosmscout/include/osmscout/async/ReadWriteLock.h]: Please add comment with motivation for custom implementation

Comment (Karry) [reply libosmscout/include/osmscout/async/ReadWriteLock.h]: Also, add `OSMSCOUT_API`, build as a shared library fails on Ubuntu:

```
[ 64%] Linking CXX executable ElevationProfileChart
/usr/bin/ld: ../libosmscout-client-qt/libosmscout_client_qtd.so.1.1.1: undefined reference to `osmscout::Latch::unlock_shared()'
/usr/bin/ld: ../libosmscout-client/libosmscout_clientd.so.1.1.1: undefined reference to `osmscout::Latch::unlock()'
/usr/bin/ld: ../libosmscout-client/libosmscout_clientd.so.1.1.1: undefined reference to `osmscout::Latch::lock()'
/usr/bin/ld: ../libosmscout-client-qt/libosmscout_client_qtd.so.1.1.1: undefined reference to `osmscout::Latch::lock_shared()'
collect2: error: ld returned 1 exit status
```

File: libosmscout/include/osmscout/async/ReadWriteLock.h [hunk 2]
```diff
@@ -24,48 +24,189 @@
...
+
+  void unlock() {
+    if (owns) {
+      owns = false;
+      p->unlock_shared();
+    }
+  }
+};
+
+class WriteLock
```
Comment (Karry) [inline libosmscout/include/osmscout/async/ReadWriteLock.h]: Please add `OSMSCOUT_API`

File: libosmscout/include/osmscout/async/ReadWriteLock.h [hunk 3]
```diff
@@ -24,48 +24,189 @@
...
+  void lock();
+  void unlock();
+
+  void lock_shared();
+  void unlock_shared();
+
+  unsigned get_shared_count() { return s_ids.size(); }
+};
+
+class ReadLock
```
Comment (Karry) [inline libosmscout/include/osmscout/async/ReadWriteLock.h]: Please add `OSMSCOUT_API`

**Review by Karry — CHANGES_REQUESTED** [review]
Hi. I just made fast review, didn't look to correctness yet.

 - Code style is different from the rest of the code, there is used camelCase, not snake_case. It is just style, but it would be great to match with the rest
 - I would suggest you to make this merge request directly to the Tim's upstream and continue with discussion there, not sure if Tim is watching this merge request
 - You wrote that you have some kind of benchmark, it would be great to add it to "Tests" directory and make it part of the test suite, to check locking correctness by Github actions on various platforms.

File: libosmscout/include/osmscout/async/ReadWriteLock.h (hunk 3 above)
Comment (janbar) [reply libosmscout/include/osmscout/async/ReadWriteLock.h]: I pushed the code relative to the benchmark (non reentrant like pthread) 1 minute ago, so later your review.
I will add more comments about your remarks ...

File: libosmscout/include/osmscout/async/ReadWriteLock.h (hunk 1 above)
Comment (janbar) [reply libosmscout/include/osmscout/async/ReadWriteLock.h]: Probably CMakeLists or meson are'nt up to date. Please try with my last push a81b48e.

Comment (janbar) [reply libosmscout/include/osmscout/async/ReadWriteLock.h]: done

File: libosmscout/include/osmscout/async/ReadWriteLock.h (hunk 2 above)
Comment (janbar) [reply libosmscout/include/osmscout/async/ReadWriteLock.h]: done

File: libosmscout/include/osmscout/async/ReadWriteLock.h (hunk 3 above)
Comment (janbar) [reply libosmscout/include/osmscout/async/ReadWriteLock.h]: done

**Comment by janbar** [comment]
I added all comments and I will open a PR for Tim later after adding testing cases.

**Comment by janbar** [comment]
The new PR for Tim is: https://github.com/Framstag/libosmscout/pull/1555

**Comment by janbar** [comment]
Hi @Karry, I tested the app in real world in navigation and I noticed the render thread keeps a read lock constantly (in move). It never releses it, i.e the count of reader never get zero, so writer cannot lock to apply style change.
The root cause is a "reentrant reader" that never drop its first lock, even if it's just a millisecond, in move. I will make some investigations ... if s_count get zero one microsec is enougth to let the writer finalizing the X lock.

OK: A reader reentrant is the root cause. I tested it by changing the line 71 and 130 to test <=1 reader to allow the writer to finalize the hold.

71:     if (s_count <= 1) {

130:   if (--s_count <= 1 && x_flag == 1 &&

And it works.

Have you an idea of which job could keep the readlock constanly in move ?

**Comment by Karry** [comment]
Wow, there is something wrong, because it was working pretty fine with pthread. I will try to reproduce it with your implementation...

**Comment by janbar** [comment]
@Karry, I found a better way to handle the case, and it works great !  I make the PR ...

Explaination: Instead locking all the time, we pass a ReadLock unlocked to jobs; then on job startup the call DBJob::run() return true or false depending if it succeeded to try_lock(). If false then the job stop immediately returning false.
This strategy works very well, and writers can hold the lock any time.
The only side effect is the logger dumps "Rendering of data has error or was interrupted" when a job is aborted. But we can update the message for this case.

**Comment by Karry** [comment]
Take a look to this commit: https://github.com/Framstag/libosmscout/commit/0c9075ddaada4f224d5efb25220cd61fd1448d88

I just added thread_local variable to count how many times one thread is holding shared lock and then debug messages for share lock/unlock method that are printing t_count, s_count, thread name and when t_count==0 it prints how long lock was locked...

I run navigation demo and grep lock log: 
```
./Demos/NavigationSimulation \
  --icons libosmscout/data/icons/svg/standard \
  --translations cmake-build-debug/OSMScout2 \
  from-home-to-work.gpx 2>&1  | grep "s_count" | tee lock.log
```

From the log, it is clear that there are two threads that are using database - map renderer and navigation. But map renderer lock is hold too long sometimes (one minute for example). During debugging, I probably find the cause: `loadJob->deleteLater()` is not called immediately after signal handler, but it is postponed somehow... So, we should close it right away and not wait for destructor. I will send merge request with the fix asap.

**Comment by janbar** [comment]
Nice !
This proves that pthread_rwlock is too optimistic (bug?) and that it was not blocking the thread writer as it should have been.
QReadWriteMutex could use pthread too underwood.

**Comment by Karry** [comment]
Please try this change: https://github.com/Framstag/libosmscout/pull/1557

**Comment by janbar** [comment]
Not fixed. It is better but 10,15 seconds to apply style change.
Applying the commit of the other way I explained above, it get 1 second max to apply style change, or show a new map downloaded. I push the commit to the PR of the Latch to show you.

**Comment by janbar** [comment]
The commit is "delegates the lock state to the task"

**Pull request closed without merging**